Use SimHits instead of digis for muon detectors in premixing

makortel opened the pull request:
This PR changes the treatment of muon detectors in premixing to use SimHits instead of SimDigis in the premixing library and in the pileup overlay. This change brings the treatment of muon detectors in premixing closer to classical mixing (where SimHits are mixed, and SimDigi producers are run with the signal+pileup SimHit input). In phase2 the SimHits were needed anyway for ME0 in addition to digis (see #23065 and https://indico.cern.ch/event/711343/#54-phase-2-premixing-status).

The effect on the event size is a slight decrease. Below is a detailed comparison for 2018 (100 TTbar events with average pileup 50)

| subdetector | IB (kB) | this PR (kB) | effect |
| --- | --- | --- | --- |
| DT | 0.67 | 5.4 | + 8x |
| CSC | 64 | 35 | - 45 % |
| RPC | 0.44 | 1.3 | + 3x |
| GEM | 0.38 | 0.53 | + 40 % |
| total | 65.5 | 41.6 | - 36 % |

The total pileup event size decreases by 0.7 % (production, i.e. without tracking MC truth). 

Since the stage where the digitization is done in changed completely, changes are expected for anything depending on muon detectors, but they should be (close to) statistical fluctuations. For GEM in phase2 workflows larger differences are seen (caused by premixing treatment becoming closer to classical mixing), see comments below for details. In fact earlier GEM was disabled for < phase2 premixing, and this PR enables it in premixing everywhere GEM is enabled.

Here are links to 100 event comparison of 250202.181 (2018 ttbar+PU FullSim) (to my private DQM GUI using an ssh tunnel recipe shown further below; black is the IB and blue this PR)
http://127.0.0.1:8081/dqm/relval/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_3_X_2018_09_26_1100-PU50_2018_premixing_100ev_orig_v1-v1/DQMIO;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=True;referenceobj1=other%3A%3A/RelValTTbar_13/CMSSW_10_3_X_2018_09_26_1100-PU50_2018_premixing_100ev_dev_v2-v1/DQMIO%3A%3A;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=M;root=;focus=;zoom=no;

10-event for 250202.171 (2017)
http://127.0.0.1:8081/dqm/relval/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_3_X_2018_09_26_1100-PU35_2017_premixing_10ev_orig_v1-v1/DQMIO;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=True;referenceobj1=other%3A%3A/RelValTTbar_13/CMSSW_10_3_X_2018_09_26_1100-PU35_2017_premixing_10ev_dev_v2-v1/DQMIO%3A%3A;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=M;root=;focus=;zoom=no;

for 250202.1 (2016)
http://127.0.0.1:8081/dqm/relval/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_3_X_2018_09_26_1100-PU35_2016_premixing_10ev_orig_v1-v1/DQMIO;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=True;referenceobj1=other%3A%3A/RelValTTbar_13/CMSSW_10_3_X_2018_09_26_1100-PU35_2016_premixing_10ev_dev_v2-v1/DQMIO%3A%3A;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=M;root=;focus=;zoom=no;

for 20234.99 (2023 D17 PU35)
http://127.0.0.1:8081/dqm/relval/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_3_X_2018_09_26_1100-PU35_2023d17_premixing_10ev_orig_v1-v1/DQMIO;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=True;referenceobj1=other%3A%3A/RelValTTbar_13/CMSSW_10_3_X_2018_09_26_1100-PU35_2023d17_premixing_10ev_dev_v2-v1/DQMIO%3A%3A;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=M;root=;focus=;zoom=no;

for 250402.1 (2016 FastSim)
http://127.0.0.1:8081/dqm/relval/start?runnr=1;dataset=/RelValTTbar_13/CMSSW_10_3_X_2018_09_26_1100-PU35_2016_fastsim_premixing_10ev_orig_v1-v1/DQMIO;sampletype=offline_relval;filter=all;referencepos=ratiooverlay;referenceshow=all;referencenorm=True;referenceobj1=other%3A%3A/RelValTTbar_13/CMSSW_10_3_X_2018_09_26_1100-PU35_2016_fastsim_premixing_10ev_orig_v1-v1/DQMIO%3A%3A;referenceobj2=none;referenceobj3=none;referenceobj4=none;search=;striptype=object;stripruns=;stripaxis=run;stripomit=none;workspace=Everything;size=M;root=;focus=;zoom=no;

with the following SSH tunnel recipe

```
ssh -L8081:mkdev7.cern.ch:8081 lxplus.cern.ch
http://127.0.0.1:8081/dqm/relval/
```

Tested in CMSSW_10_3_X_2018-09-26-1100 (rebased on top of 10_3_0_pre6), expecting changes in muon detectors etc. in premxing workflows. Because the premixing library format changes (for muons), this PR breaks all stage2-only premixing workflows.

@kpedro88 @mdhildreth

makortel committed:
Use SimHits instead of digis for muons

This is closer to what is done in classical mixing, and turns out to
lead slightly smaller file sizes as well.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24827/6777

cmsbuild commented:
A new Pull Request was created by  @makortel (Matti Kortelainen) for master.

It involves the following packages:

Configuration/StandardSequences
FastSimulation/Configuration
RecoLocalMuon/RPCRecHit
SimGeneral/MixingModule
SimGeneral/PreMixingModule
SimMuon/CSCDigitizer
SimMuon/Configuration
SimMuon/DTDigitizer
SimMuon/GEMDigitizer
SimMuon/MCTruth
SimMuon/RPCDigitizer
Validation/GlobalDigis
Validation/MuonCSCDigis
Validation/MuonDTDigis
Validation/MuonRPCDigis


@perrotta, @andrius-k, @lveldere, @kmaeshima, @civanch, @schneiml, @ssekmen, @mdhildreth, @cmsbuild, @franzoni, @jfernan2, @kpedro88, @slava77, @fabiocos, @davidlange6 can you please review it and eventually sign? Thanks.
@trocino, @battibass, @felicepantaleo, @abbiendi, @GiacomoSguazzoni, @jhgoh, @VinInn, @Martin-Grunewald, @calderona, @HuguesBrun, @drkovalskyi, @dildick, @matt-komm, @bellan, @ptcox, @rovere, @dgulhan, @folguera this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test workflow 250202.181,250202.171,250202.1,20234.99,250402.1

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: b9ff4a45d04370e4ef300afe8e04b81559105af3
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24827/30929/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24827/30929/summary.html

@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-24827/20234.99_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D17PU_GenSimHLBeamSpotFull14+SingleNuE10_cf_2023D17PU_PremixHLBeamSpotFull14PU_Premix+DigiFullTriggerPUPRMXCombined_2023D17PU+RecoFullGlobalPUPRMX_2023D17PU+HARVESTFullGlobalPU_2023D17PU
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-24827/250202.1_TTbar_13+TTbar_13+PREMIXUP15_PU25+DIGIPRMXLOCALUP15_PU25+RECOPRMXUP15_PU25+HARVESTUP15_PU25
   * /build/cmsbld/jenkins/workspace/compare-root-files-short-matrix/results/JR-comparison/PR-24827/250402.1_TTbar_13+FS_PREMIXUP15_PU25+FS_TTbar_13_PRMXLOCALUP15_PU25+HARVESTUP15FS

Comparison Summary:
 - No significant changes to the logs found
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 576 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3162979
 - DQMHistoTests: Total failures: 6163
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3156619
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 134 log files, 14 edm output root files, 32 DQM output files

drkovalskyi commented:
Sorry for off topic, is there any chance we can use the SimHits from premixed library to get the TrackingPartickes for muons from pileup?

makortel commented:
@drkovalskyi IIRC TrackingParticles are created from SimTracks (in `TrackingTruthAccumulator`. I'm not sure if one could hack through with SimHits. In addition, in this PR SimHits are included only for muon detectors, so all information would be limited to muon detectors.

slava77 commented:
In jenkins tests I see changes only in 250202.181, the 2018 wf with PU and premixing. This is expected.
In general there is just some shuffle in the distributions.
In one case there is an appearance (from empty to non-empty content): this is happening for GEM hits and segments
![all_oldvsnew_ttbar13tevpuppmx2018wf250202p181c_gemdetidgemsegmentsownedrangemap_gemsegments__reco_obj_collection__data_at_size](https://user-images.githubusercontent.com/4676718/46686450-01153200-cc01-11e8-87af-0366905cfd4d.png)

<img width="742" alt="wf250202 181_gemdigi_deta" src="https://user-images.githubusercontent.com/4676718/46687348-5fdbab00-cc03-11e8-823d-a7429bd120e2.png">

Is this a fluctuation of some noise or did something change in a significant way?

makortel commented:
Thanks @slava77. On closer look I noticed that these customizations
https://github.com/cms-sw/cmssw/blob/b9ff4a45d04370e4ef300afe8e04b81559105af3/Validation/MuonGEMDigis/python/MuonGEMDigis_cff.py#L55-L58
have to be removed, even though they have no effect on 2018 workflow. On the other hand, what could be going on is that those customizations should have been enabled in the IB, i.e. currently the module is looking for wrong inputs, while this PR makes those inputs correct, and therefore contents show up in DQM.

I'd expect the change in `gemSegments` size to be a fluctuation, but I'll check it again with 100 events (the IB has been removed so I have to rebase so it will take a while).

makortel commented:
@kpedro88 In IB yes (even though it could be simplified to just `premix_stage2`, assuming no-one would run GEM validation without GEM era). In this PR they should be removed.

Assuming the review will (again) take a while, would it be worth to submit a PR for interim fx?

slava77 commented:
> Assuming the review will (again) take a while, would it be worth to submit a PR for interim fx?

given zero signatures in this PR now; was the question in the context of the following?
1. update this PR with a fix
1. make a side PR with a fix and wait here until that one is merged

kpedro88 commented:
I think the fix would have to be merged into the IB first, in order to provide a better comparison for this PR.

makortel commented:
Indeed the fix makes sense only for IB (so Slava's option 2), as for 2018 this PR is already correct (but 2023 needs to be fixed in this PR). I'll prepare it.

makortel commented:
With 100 events, comparing classical mixing (left) to premixing (right) in IB gives this
![image](https://user-images.githubusercontent.com/33993/46974531-dfe19500-d089-11e8-8221-dd65e4a2cb4d.png)

and comparing classical mixing (left) to premixing (right) with this PR
![image](https://user-images.githubusercontent.com/33993/46974561-f7b91900-d089-11e8-952d-4151a4aaeee6.png)

so (by quick look) this PR appears indeed model GEM better in premixing than the version in IB.

kpedro88 commented:
Looks like a clear improvement to me. Will you update the PR to remove the unwanted customizations in `Validation/MuonGEMDigis/python/MuonGEMDigis_cff.py`?

makortel committed:
Fix GEM validation for phase2 premixing

These parameters point to "real digis" instead of "sim digis", so they
should not be customized for premixing.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-24827/6865

makortel commented:
@kpedro88 Done (also rebased on top of 10_3_0_pre6).

makortel commented:
@cmsbuild, please test workflow 250202.181,250202.171,250202.1,20234.99,250402.1

cmsbuild commented:
Pull request #24827 was updated. @perrotta, @andrius-k, @lveldere, @kmaeshima, @civanch, @schneiml, @ssekmen, @mdhildreth, @cmsbuild, @franzoni, @jfernan2, @kpedro88, @slava77, @fabiocos, @davidlange6 can you please check and sign again.

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 3299fc3689c4323942a9be4a9e7e2c88d53f0d7a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-24827/31089/summary.html

cmsbuild commented:
Comparison job queued.

slava77 commented on SimGeneral/PreMixingModule/python/mixOne_premix_on_sim_cfi.py:
```diff
@@ -199,10 +163,39 @@
+from Configuration.Eras.Modifier_run2_GEM_2017_cff import run2_GEM_2017
+from Configuration.Eras.Modifier_run3_GEM_cff import run3_GEM
+(run2_GEM_2017 | run3_GEM).toModify(
+    mixData,
```
it looks like that this update is the main reason the GEMs now show up in the mixed outputs in amount more consistent with the classic mixing.

makortel replied:
Correct, I'll add that to the PR description (thanks for pointing it out).

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/24827	3299fc3
- some infrastructure changes mostly unrelated to reco. One reco file update is RecoLocalMuon/RPCRecHit/python/rpcRecHits_cfi.py  is for phase2 premixing setup to use default inputs.
    - apparently this PR includes a bugfix for GEM premixing for run2: now the GEM hits are showing up at about the same rate as the classic mixing case
- jenkins tests pass and comparisons with the baseline show differences only in the 2018 premix workflow 250202.181 with changes starting from muon hits.

kpedro88 commented:
+upgrade

ssekmen commented:
+1

civanch commented:
+1

fabiocos commented:
@jfernan2 @andrius-k could you please check? It is mostly a cleaning of validation fragments, as far as DQM is concerned

andrius-k commented:
+1

fabiocos commented:
+operations

the changes to StandardSequences look coherent with the purpose of the PR

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
@smuzaffar this PR should again break a number of test workflows until the new input files are produced. I keep its merge for the last IB before building 10_4_0_pre1 to minimize the impact on all the rest

fabiocos commented:
+1
